Add comment about Thanos, external labels and downsampling. (#2122)
grafana/mimir: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md

gsutil -m cp -r <DIRECTORY> gs://<TENANT>/<DIRECTORY>
```

> ## Block metadata

Author: I get what you're saying. I think the answer right now depends on your needs:
- if you don't mind losing external labels in results, you can upload blocks as-is, and forget about them
- if you used deduplication, you most likely only want to upload blocks from one replica, but not the other
- if you DO mind losing external labels, currently we don't have a solution. We will one partial solution in backfilling API – writing external labels into series, if they don't conflict with existing labels.

Shall we be explicit about this?

> ## Block metadata

Reviewer: Yes. I think that's exactly what you should document in a user-facing doc, not the reason behind it, the difference between versions and so on. Keep in mind doc is versioned, so I think we should just address 2.2.0.

> ## Block metadata

Author: I have introduced `## Considerations on Thanos specific features` section and completely rephrased it to be explicit about what we recommend in different scenarios. Please take a look again.

> ## Block metadata

Reviewer 2: Hi, just passing by, thinking about how to do migration from Thanos. 

Few points:
 - If you want to preserve external labels, you should in theory be able to use the `thanos bucket rewrite --rewrite.to-relabel-config="..."` to rewrite all the blocks and add te external labels to it. Thah could work I hope? See https://thanos.io/tip/components/tools.md/#bucket-rewrite
 - And about deduplication deduplication, there is possibility to use vertical compaction with deduplication first, using something like  `thanos compact --compact.enable-vertical-compaction --deduplication.func=penalty --deduplication.replica-label=LABEL` see https://thanos.io/tip/components/compact.md/#vertical-compaction-use-cases 
 
Might be worth adding note about those?

> ## Block metadata

Author: Thank you @FUSAKLA! That's very useful, I will add it to the PR.

gsutil -m cp -r <DIRECTORY> gs://<TENANT>/<DIRECTORY>
```

## Block metadata

Every block has a `meta.json` metadata file used by Grafana Mimir, Prometheus, and Thanos to identify the block contents.
Each project has its own metadata conventions.

> Ingesters from Mimir 2.1 and earlier added an external label into `meta.json` file to identify the tenant that owns the block.

Reviewer 3: ```suggestion
From the Grafana Mimir 2.1 release or earlier, the ingesters added an external label into the `meta.json` file to identify the tenant that owns the block.
```

Reviewer 3: This information sounds like information that ought to be in the release notes rather than the core documentation.

Author: It is mentioned in the Mimir 2.2 release notes as:

```
* [CHANGE] Blocks uploaded by ingester no longer contain `__org_id__` label. Compactor now ignores this label and will compact blocks with and without this label together. `mimirconvert` tool will remove the label from blocks as "unknown" label. #1972
```

> Mimir compactor also adds special label identifying compactor-shard when split-and-merge sharding feature is enabled.

Reviewer 3: @pstibrany, I assume a human is performing this action. Passive voice makes that ambiguous. Please vet for accuracy.

> Mimir compactor also adds special label identifying compactor-shard when split-and-merge sharding feature is enabled.

Author: Thanks for suggestion. I would rephrase it slightly to associate "sharding feature" with "split-and-merge compactor".

```suggestion
If you enable the sharding feature of Mimir's split-and-merge compactor, compactor adds a special label to identify the compactor-shard.
```

WDYT?

> In Mimir 2.2 and later blocks no longer have a label identifying the tenant. Compactor shard label is still used.

Reviewer 3: This information sounds like information that ought to be in the release notes rather than the core documentation.

> In Mimir 2.2 and later blocks no longer have a label identifying the tenant. Compactor shard label is still used.

Author: > What technical value does "still" have here?

It is the only label left that Mimir (still) uses internally. All other internal labels are not used by Mimir anymore.

> When querying blocks, Mimir ignores all external labels and doesn't add them to the results.

Reviewer 3: ```suggestion
When querying blocks, Mimir ignores all external labels and doesn't add them to the results.
```
What does "When querying blocks" mean?
* Does it get blocked (by what?)
* Does it block something? In either case, who or what, and by whom?

Author: > What does "When querying blocks" mean?
> 
> * Does it get blocked (by what?)
> * Does it block something? In either case, who or what, and by whom?

"blocks" is used as a noun here, not verb. When user sends a query to execute, we say that "Mimir queries (TSDB) blocks".

> During compaction Mimir will respect external labels in the block and only merge blocks with the same labels together.

Reviewer 3: ```suggestion
During compaction, Mimir respects the external labels in the block and only merges blocks with the same labels.
```

> Note that blocks from Prometheus do not have any labels stored in them, only blocks from Thanos use them.

Reviewer 4: (not sure if my suggestion is correct, but if this is what you meant I think it's less ambiguous)
```suggestion
Note that blocks from Prometheus do not have any external labels stored in them, only blocks from Thanos use them.
```

> Note that blocks from Prometheus do not have any external labels stored in them, only blocks from Thanos use them.

Reviewer 3: ```suggestion
> **Note:** Blocks from Prometheus do not have any external labels stored in them; only blocks from Thanos use labels.
```

> ### Note about Thanos: external labels and downsampling

Reviewer 3: ```suggestion
```

> Thanos requires that Prometheus is configured with external labels.
> When Thanos sidecar uploads blocks, it includes these external labels into `meta.json` file inside the block.
> Thanos then returns these external labels when block is queried, and can also use labels for deduplication of replicated data.

Reviewer 3: ```suggestion
> **Note:** Thanos requires that Prometheus is configured with external labels.
When the Thanos sidecar uploads blocks, it includes the external labels from Prometheus in the `meta.json` file inside the block.
When you query the block, Thanos returns Prometheus’ external labels. Thanos also uses labels for the deduplication of replicated data.
```
As a fellow human, am I querying the block or is something doing that for me?

Mimir doesn't include external labels in the query results.

> This means that blocks that were originally created by Thanos will miss all of their external labels in the results when queried.

Reviewer 4: ```suggestion
This means that blocks that were originally created by Thanos will not include their external labels in the results when queried.
```

> Mimir doesn't include external labels in the query results.
> This means that blocks that were originally created by Thanos will not include their external labels in the results when queried.

Reviewer 3: ```suggestion
Blocks that were originally created by Thanos do not include their external labels in the query results, because Mimir doesn't include external labels in the query results.
```

Author: Without "Mimir doesn't include external labels in the query results.", the sentence doesn't mention that this happens when Mimir uses these blocks from Thanos.

Author: But since previous paragraph is a "note about thanos", then it shows that this one is about Mimir. I think that works.

> Thanos can also be configured to use labels for deduplication. After uploading such blocks into Mimir, query results will contain duplicate samples (assuming they don't use exact same timestamp).
> Mimir's Compactor will however not compact such blocks together, as long as blocks have different set of external labels.

Reviewer 3: ```suggestion
You can configure Thanos to use labels for deduplication. After uploading such blocks into Mimir, the query results will contain duplicate samples (assuming their timestamps are not identical).
As long as the blocks have different set of external labels, Mimir's compactor does not compact such blocks together.
```

Mimir also doesn't understand Thanos "downsampling" feature, that is `Resolution` field from `meta.json` file created by Thanos is ignored by Mimir.

> Notable exception is a compactor, which will not compact blocks with different resolutions together.

Reviewer 4: ```suggestion
A notable exception is the Mimir compactor, which will not compact blocks with different resolutions together.
```

> Mimir also doesn't understand Thanos "downsampling" feature, that is `Resolution` field from `meta.json` file created by Thanos is ignored by Mimir.
> A notable exception is the Mimir compactor, which will not compact blocks with different resolutions together.
> Blocks generated by Mimir don't set the resolution field at all.

Reviewer 3: ```suggestion
> **Note:** Mimir does not understand Thanos’ _downsampling_ feature. As a result, Mimir ignore the `Resolution` field from the `meta.json` file that Thanos creates.
A notable exception is that the Mimir compactor does not compact together blocks of different resolutions.
The blocks that Mimir generates have unset resolution fields.
```

gsutil -m cp -r <DIRECTORY> gs://<TENANT>/<DIRECTORY>
```

## Block metadata

Each block has a `meta.json` metadata file that is used by Grafana Mimir, Prometheus, and Thanos to identify the block contents.
Each project has its own metadata conventions.

In the Grafana Mimir 2.1 (or earlier) release, the ingesters added an external label to the `meta.json` file to identify the tenant that owns the block.

> If you enable the sharding feature of Mimir's split-and-merge compactor, compactor adds a special label to identify the compactor-shard.

Reviewer: This sentence looks like related to 2.1 or earlier only.

Author: Next paragraph says it still applies for 2.2:

> The compactor-shard label is still used, if sharding of blocks during compaction is enabled.

Reviewer: I know but people sometimes just stop reading. Given the split-and-merge compactor shard label doesn't change by Mimir versions, I wouldn't mention it along with 2.1 or 2.2 changes at all.

Author: > Given the split-and-merge compactor shard label doesn't change by Mimir versions, I wouldn't mention it along with 2.1 or 2.2 changes at all.

OK.

In the Grafana Mimir 2.2 (or later) release, blocks no longer have a label that identifies the tenant. The compactor-shard label is still used, if sharding of blocks during compaction is enabled.

> **Note**: Blocks from Prometheus do not have any external labels stored in them; only blocks from Thanos use labels.

## Considerations on Thanos specific features

> > **Note**: Thanos requires that Prometheus is configured with external labels.

Reviewer: Looks a bit weird to me starting the section with a "note" (so far I've seen notes being addendum at the end of a section). I would consider making these sentences normal ones, not a "note".

> When the Thanos sidecar uploads blocks, it includes the external labels from Prometheus in the `meta.json` file inside the block.
> When you query the block, Thanos injects Prometheus’ external labels in the series returned in the query result. Thanos also uses labels for the deduplication of replicated data.

If you want to use existing blocks from Thanos by Grafana Mimir, there are some considerations:

**Grafana Mimir doesn't inject external labels into query results.**
This means that blocks that were originally created by Thanos will not include their external labels in the results when queried.
If you need to have external labels in your query results, this is currently not possible to achieve in Grafana Mimir.

**Grafana Mimir will not respect deduplication labels configured in Thanos when querying the blocks.**

> For the best results please only upload Thanos blocks from single Prometheus replica from each HA pair.

Reviewer: What does "best results" mean here? If I upload all Thanos blocks, are query results incorrect or just less performant? I think they're just less performant, so I would clarify it ("For the best query performance...").

Author: Good point.

If you upload blocks from both replicas, the query results returned by Mimir will include samples from both replicas.

**Grafana Mimir does not support Thanos’ _downsampling_ feature.**

> For best results please only upload original (raw) Thanos blocks into Mimir's storage.

Reviewer: In this case "best results" means "To guarantee query results correctness". I would clarify it.

Author: Clarified.

Every block has a `meta.json` metadata file used by Grafana Mimir, Prometheus, and Thanos to identify the block contents.
Each project has its own metadata conventions.
The `metaconvert` tool updates the `meta.json` metadata for use in Mimir.

In Mimir 2.1 and earlier, `metaconvert` added external label into `meta.json` file to identify the tenant that owns the block.
It also removed all other labels, only preserving labels known to Mimir.

In Mimir 2.2 and later blocks no longer have a label identifying the tenant, and `metaconvert` simply removes all external labels from `meta.json` file not known to Mimir.

Mimir itself does not use labels from `meta.json` file. Note that blocks from Prometheus do not have any labels stored in them,
only blocks from Thanos use them.

> ### Note about Thanos: external labels and downsampling

Author: Now it's:

```
> **Note:** Mimir does not support Thanos’ _downsampling_ feature. As a result, Mimir ignores the `Resolution` field from the `meta.json` file that Thanos creates.
```

Thanos requires that Prometheus is configured with external labels.
When Thanos sidecar uploads blocks, it includes these external labels into `meta.json` file inside the block.
Thanos then returns these external labels when block is queried, and can also use labels for deduplication of replicated data.

Mimir doesn't support external labels feature, and only uses few labels internally. `metaconvert` will remove all external labels from `meta.json` file, and only keep labels used by Mimir, if they are valid.

This means that blocks that were originally created by Thanos will miss some of the labels when queried.

If Thanos system was using labels for deduplication, and all such blocks are uploaded to Mimir's bucket, after running `metaconvert` and removing the labels, Mimir will compact such blocks together.
This will result in blocks with duplicated samples (unless sample timestamps were exactly the same).

Mimir also doesn't understand Thanos "downsampling" feature.
While `metaconvert` currently doesn't remove `Resolution` field from `meta.json` file, the rest of Mimir does not respect the field during query time.

> Notable exception is a compactor, which will not compact blocks with different resolutions together.

Author: Based on ealier feedback, this now reads:

```
A notable exception is the Mimir compactor, which will not compact blocks with different resolutions together.
```

WDYT?
